File: src/control/OptimizeController.java
package control;

import model.*;

import java.sql.Date;
import java.sql.Time;
import java.util.ArrayList;
import java.util.List;

/**
 * OptimizeController
 * Handles all functionality for the use-case optimize.
 *
 * @author Dani Sander
 * @version 1.0
 * @since 27-05-15
 */

public class OptimizeController {
    private RouteController routeController;
    private MapController mapController;
    private static OptimizeController instance;
    private double load;
    private ArrayList<DeliveryStop> removedStops;
    private boolean isnear;
    private double bestdistance;
    private Route bestRoute;
    
    /**
     * Private constructor for singleton.
     */
    private OptimizeController() {
        
        routeController = RouteController.getInstance();
        mapController = MapController.getInstance();
        removedStops = new ArrayList<>();
    }
    
    /**
     * Singleton method for class.
     * @return instance of class.
     */
    public static OptimizeController getInstance() {
        if (instance == null) {
            instance = new OptimizeController();            
        }
        
        return instance;
    }
    
    
    /**
     * Optimizes all imported routes.
     */
    public void optimize() {
        
        //Finds all needed routes.
        ArrayList<Route> overloadedRoutes = routeController.findOverloadedRoutes();
        ArrayList<Route> underloadedRoutes = routeController.findUnderloadedRoutes();
        List<Route> allRoutes = routeController.getRoutes();
        
        //Loads maps.
        mapController.loadMaps(allRoutes);
            
        //Checks to see if there is overloadedRoutes.
        if (overloadedRoutes.size() >= 1) {
            
            //Enters a loop for each overloaded route.
            overloadedRoutes.stream().forEach((overloadedRoute) -> {
                
                //Finds overloaded amount.
                double overload = findOverloadAmount(overloadedRoute);
                
                //Removes deliveryStops from route, until it's not overloaded, using a greedy algorithm.
                while (overload > 0) {
                    
                    //Finds the most/best overloaded stop.
                    DeliveryStop best = findBestOverloadedStop(overloadedRoute, overload);
                    
                    //Removes stop from route.
                    overloadedRoute.getStops().remove(best);
                    
                    //Adds stop to ArrayList.
                    removedStops.add(best);
                    
                    //Decrements overload.
                    overload -= findLoad(best);
                }
                
                //Checks to see if there is underloadedRoutes.
                if (underloadedRoutes.size() >= 1) {
                    
                    //ArrayList containing all nearRoutes.
                    ArrayList<Route> nearRoutes = new ArrayList<>();
                    
                    //Enter a loop for each stop removed.
                    removedStops.stream().forEach((removedStop) -> {
                        
                        //Finds the geoLoc for current removedStop.
                        GeoLoc geoLocRemovedStop = mapController.findGeoLoc(removedStop);
                        
                        //Enters a loop for each underloadedRoute, to check if one of them is near stop.
                        underloadedRoutes.stream().forEach((underloadedRoute) -> {
                            
                            //Misc Variables for the loop.
                            boolean isnear_route = false;
                            double bestdistance = 9999;
                            
                            //Enters a loop for each stop.
                            ArrayList<DeliveryStop> stops = underloadedRoute.getStops();
                            for(DeliveryStop deliveryStop:stops) {
                                
                                //Gets the geoLoc, and checks to see if one of those are near the removed stop.
                                GeoLoc geoLoc = mapController.findGeoLoc(deliveryStop);
                                double distance = mapController.pointDistance(geoLoc, geoLocRemovedStop);
                                
                                //Checks to see if within a certain distance.
                                if (distance < 25) {
                                    
                                    //Checks to see if shorter then best distance.
                                    if (distance < bestdistance) {
                                        
                                        //Sets to the new best distance.
                                        bestdistance = distance;
                                        isnear_route = true;
                                        isnear = true;
                                    }
                                }
                            }
                        
                            //Adds the underloadedRoute to ArrayList of nearRoutes.
                            if (isnear_route) {
                                nearRoutes.add(underloadedRoute);
                            }
                        });
                        
                        if (isnear) {
                            
                            //Checks to see if more than one are near the stop.
                            int nearRouteQuantity = nearRoutes.size();
                            if (nearRouteQuantity > 1) {
                                
                                //Best distance variable.
                                bestdistance = 9999;
                                bestRoute = nearRoutes.get(0);
                                
                                //Enters a loop for each route to find the best.
                                nearRoutes.stream().forEach((nearRoute) -> {
                                    
                                    //Enters a loop for each stop.                                
                                    ArrayList<DeliveryStop> stops = nearRoute.getStops();
                                    for(DeliveryStop deliveryStop:stops) {
                                        
                                        //Gets the geoLoc and distance of the current stop.
                                        GeoLoc geoLoc = mapController.findGeoLoc(deliveryStop);
                                        double distance = mapController.pointDistance(geoLoc, geoLocRemovedStop);
                                        
                                        //Checks to see if better then best distance.
                                        if (distance < bestdistance) {
                                            bestdistance = distance;
                                            bestRoute = nearRoute;
                                        }
                                    }
                                });

                                //Move stop to this route.
                                Route underloadedRoute = nearRoutes.get(0);
                                underloadedRoute.addDeliveryStop(removedStop);
                            }
                            else {
                                
                                //Move stop to this route.
                                Route underloadedRoute = nearRoutes.get(0);
                                underloadedRoute.addDeliveryStop(removedStop);
                            }
                        }
                        else {
                            
                            //Make a new route.
                            Route newRoute = newRoute(removedStops);
                            underloadedRoutes.add(newRoute);
                            allRoutes.add(newRoute);
                        }
                    });
                }
                else {
                    
                    //Make a new route.
                    Route newRoute = newRoute(removedStops);
                    underloadedRoutes.add(newRoute);
                    allRoutes.add(newRoute);
                }
            });  
            
            //Updates all routes again.
            routeController.setRoutes(allRoutes);
            
            //Clears ArrayList
            removedStops.clear();
    }
}
	
	
	/**
	 * Makes a new route with the given delivery stops.
	 * @param deliveryStops
	 * @return returns the route made.
	 */
	private Route newRoute(ArrayList<DeliveryStop> deliveryStops) {
		
		//Create default route.
		Time timeOfDeparture = new Time(3, 0, 0);
		TrailerType trailerType = TrailerType.STOR;
		boolean extraRoute = true;
		
		DefaultRoute defaultRoute = new DefaultRoute(timeOfDeparture, trailerType, extraRoute);
		
		//Add default stops.
		for(DeliveryStop deliveryStop:deliveryStops) {
			
			//Gets the default stop from normal stop and adds it to defaultRoute.
			DefaultDeliveryStop defaultStop = deliveryStop.getDefaultStop();
			defaultRoute.addDefaultDeliveryStop(defaultStop);	
		}
		
		//Create route.
		Date date = new Date(0, 0, 0);
		Route route = new Route(defaultRoute, date);
		
		//Add stops.
		route.setStops(deliveryStops);
		
		//Return route.
		return route;
	}
	
	
	
	/**
	 * Finds the amount overloaded for the route.
	 * @param route route to find overload amount for.
	 * @return the amount overloaded in double.
	 */
	private double findOverloadAmount(Route route) {
		
		//Variable to increment for each load check.
    	load = 0;
    	
    	//Finds maximum load.
    	double capacity = route.getDefaultRoute().getTrailerType().getCapacity();       	
    	
    	//Enters a loop for each delivery stop.
    	ArrayList<DeliveryStop> stops = route.getStops();       	
    	stops.stream().forEach((stop) -> {
    		
    		//Enters a loop for each transportUnit
    		ArrayList<TransportUnit> transportUnits = stop.getTransportUnits();
    		for(TransportUnit transportUnit:transportUnits) {
    			
    			//Increments load with the transportUnits size.
    			load += transportUnit.getUnitType().getSize();
    		}
    	});
    	
    	//Finds amount overloaded, and returns that.
    	double overload = load - capacity;
		return overload;
	}
	
	
	
	/**
	 * Finds the most overloaded delivery stop in the given route, as long as it doesn't exceed the overload cap.
	 * If no such stop can be found, finds the least overloaded route instead.
	 * @param route the route whose stops are to be checked.
	 * @param overload the overload cap to exclude stops whose load are not optimal.
	 * @return the best suited deliveryStop whose load most closely fits the cap.
	 */
	private DeliveryStop findBestOverloadedStop(Route route, double overload) {
		
		//Finds an initial stops for comparison.
		DeliveryStop biggest = route.getStops().get(0);
		DeliveryStop best = biggest;
		
		//Boolean to check if cap is exceeded.
		boolean exceedCap = false;
		
		//Enters a loop for each stop, to find the most overloaded stop.
		ArrayList<DeliveryStop> stops = route.getStops();
		for(DeliveryStop deliveryStop:stops) {
			
			//Compares the load of the deliveryStop with the load of the biggest, to find which is biggest.
			double biggestload = findLoad(biggest);
			double compareload = findLoad(deliveryStop);
			if (compareload > biggestload) {
				
				//Sets the biggest to the current.
				biggest = deliveryStop;
				best = biggest;
				
				//Checks to see if the biggest exceeds the cap.
				if (biggestload > overload) {
					exceedCap = true;
				}
				else{
					exceedCap = false;
				}
			}
		}
		
		//Checks to see if the biggest exceeds the limit.
		if (exceedCap) {
			
			//Initial stop for comparison.
			DeliveryStop smallest = route.getStops().get(0);
			
			//Enters a loop for each stop, to find the least overloaded stop.
			for(DeliveryStop deliveryStop:stops) {
				
				//Compares the load of the deliveryStop with the load of the smallest, to find which is smallest.
				double smallestload = findLoad(smallest);
				double compareload = findLoad(deliveryStop);
				if (compareload < smallestload) {
					
					//Sets the smallest to current.
					smallest = deliveryStop;
					best = smallest;
				}
			}	
		}
		
		//Returns the best load.
		return best;
	}
	
	
	/**
	 * Finds the load of the given delivery stop.
	 * @param stop stop to find load of.
	 * @return the load in double.
	 */
	private double findLoad(DeliveryStop stop) {
		
		//Creates variable.
		double load = 0;
		
		//Enters a loop for each transportUnit
		ArrayList<TransportUnit> transportUnits = stop.getTransportUnits();
		for(TransportUnit transportUnit:transportUnits) {
			
			//Increments load with the transportUnits size.
			load += transportUnit.getUnitType().getSize();
		}
		
		//Returns the load.
		return load;
	}
}

File: Control/OptimizeController.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using GMap.NET;
using Model;

namespace Control
{
    public class OptimizeController
    {
        public RouteController RouteCtr  { get; private set; }
        public MapController MapCtr { get; private set; }
        public LogController LogCtr { get; private set; }
        private static OptimizeController instance;

        private ConcurrentQueue<DeliveryStop> removedStops;
        private List<Route> overloadedRoutes;
        private List<Route> underloadedRoutes;
        private int count1;
        private int count2;

        // /<summary>
        // /Private singleton constructor.
        // /</summary>
        private OptimizeController() {
            RouteCtr = RouteController.Instance;
            MapCtr = MapController.Instance;
            LogCtr = LogController.Instance;
        }

        // /<summary>
        // /Singleton method. Returns the Instance of the class.
        // /</summary>
        // /<returns>Instance of class.</returns>
        public static OptimizeController Instance {
            get
            {
                if (instance == null)
                {
                    instance = new OptimizeController();
                }
                return instance;
            }
        }

        /**
         * Optimizes all imported routes.
         */
        public void Optimize() {

            count1 = 0;
            count2 = 0;

            // Sync Queue
            removedStops = new ConcurrentQueue<DeliveryStop>();

            LogCtr.StatusLog("searching for over- and underloaded");
            overloadedRoutes = RouteCtr.FindOverloadedRoutes();
            underloadedRoutes = RouteCtr.FindUnderloadedRoutes();

            LogCtr.StatusLog("Found all over- and under-loaded routes");

            // Checks to see if there is overloadedRoutes.
            if (overloadedRoutes.Count > 0) {

                // Enters a loop for each overloaded route.
                Parallel.ForEach(overloadedRoutes, overloadedRoute => {

                    // Removes deliveryStops from route, until it's no longer overloaded, using a greedy algorithm.
                    List<DeliveryStop> stops = FindAndRemoveStops(overloadedRoute, RouteCtr.Routes);
                    foreach (DeliveryStop stop in stops)
                    {
                        removedStops.Enqueue(stop);
                    }

                    count1++;
                });
                LogCtr.StatusLog(String.Format("There is %d delivery stop there need to be moved to another routes", overloadedRoutes.Count));

                // Enter a loop for each stop removed.
                foreach (DeliveryStop removedStop in removedStops)
                {
                    // The thread-safe dictionary contains all routes near removedStop with distance as the key.
                    ConcurrentDictionary<Double, Route> nearRoutes = new ConcurrentDictionary<double, Route>();

                    LogCtr.StatusLog("Searching for routes at a close enough distance, and with enough free space to add the stop");
                    // Enters a loop for each underloadedRoute, to check if one of them is near stop.
                    Parallel.ForEach(underloadedRoutes, underloadedRoute => {  // TODO: make parallelStream
                        // Checks if there is space to removedStop on this Route
                        double newLoad = removedStop.GetSizeOfTransportUnits() + underloadedRoute.GetLoadForTrailer();
                        if (newLoad <= underloadedRoute.GetCapacity()) {
                            // Finds the shortest distance from removedStop to underloadedRoute.
                            MapRoute mapRoute = MapController.ShortestDistancesTo(removedStop, underloadedRoute);
                            
                            // Checks to see if within a acceptable distance.
                            if (mapRoute.Distance < 25) {
                                nearRoutes.AddOrUpdate(mapRoute.Distance, underloadedRoute, (d, route) => route);
                            }
                        }
                    });

                    if (nearRoutes.Count > 0) {
                        // The shortest distance (HashMap key) for all the Routes
                        double distance = nearRoutes.Keys.Min();

                        // Move stop to this route.
                        Route underloadedRoute = nearRoutes[distance];
                        underloadedRoute.Stops.Add(removedStop);
                        if (!underloadedRoute.IsUnderloaded()) {
                            underloadedRoutes.Remove(underloadedRoute);
                        }

                        LogCtr.StatusLog("Added delivery stop to route");
                    } else {
                        // Make a extra route.
                        Route extraRoute = ExtraRoute(removedStop);
                        RouteCtr.Routes.Add(extraRoute);
                        if (extraRoute.IsUnderloaded()) {
                            underloadedRoutes.Add(extraRoute);
                        }

                        LogCtr.StatusLog("Created a new route for delivery stop");
                    }

                    count2++;
                }

                RouteCtr.CalcTimeForDeparture();
            }

            //Parallel.ForEach(RouteCtr.Routes, route => MapCtr.PreCalcRoad(route));
            foreach (Route route in RouteCtr.Routes)
            {
                MapCtr.PreCalcRoad(route);
            }
        }


        /**
         * Makes a new route and add a delivery stop.
         *
         * @param deliveryStop
         * @return returns the new route.
         */
        private static Route ExtraRoute(DeliveryStop deliveryStop) {
            // Create default route.
            DefaultRoute defaultRoute = new DefaultRoute(trailerType: TrailerType.STOR, extraRoute: true);

            // Create route.
            Route route = new Route(defaultRoute, DateTime.Now);

            // Add stops.
            route.Stops.Add(deliveryStop);

            // Return extra route.
            return route;
        }


        /**
         * Finds the most overloaded delivery stop in the given route, as long as it doesn't exceed the overload cap.
         * If no such stop can be found, finds the least overloaded route instead.
         *
         * @param route    the route whose stops are to be checked.
         * @param overload the overload cap to exclude stops whose load are not optimal.
         * @return the best suited deliveryStop whose load most closely fits the cap.
         */

        public static List<DeliveryStop> FindAndRemoveStops(Route route, ConcurrentBag<Route> Routes)
        {
            List<DeliveryStop> removedStops = new List<DeliveryStop>();

            // Sort delivery stops after its load size, with the biggeste load first
            DeliveryStop[] stops = route.Stops.ToArray();
            Array.Sort(stops, (stop1, stop2) =>
            {
                if (stop1.GetSizeOfTransportUnits() < stop2.GetSizeOfTransportUnits())
                    return 1;
                if (stop1.GetSizeOfTransportUnits() > stop2.GetSizeOfTransportUnits())
                    return -1;
                return 0;
            });

            // Removes all delivery stops from route
            route.Stops.Clear();

            for(int x = 0; x < stops.Length; )
            {
                // Check if the size of the trasnsport units is more then there is space to in the trailer
                if (route.DefaultRoute.TrailerType >= stops[x].GetSizeOfTransportUnits())
                {
                    // Checks if there is enough space in the Trailer else adds the stop to removedStops
                    if (route.GetCapacity() >= route.GetLoadForTrailer() + stops[x].GetSizeOfTransportUnits())
                    {
                        route.Stops.Add(stops[x]);
                    }
                    else
                    {
                        removedStops.Add(stops[x]);
                    }
                    x += 1;
                }
                else // Create a extra route for this stop, becasuse there is not enough space in the trailer
                {
                    DeliveryStop newStop = new DeliveryStop(stops[x].DefaultStop);
                    Route extraRoute = ExtraRoute(newStop);

                    // Sort transport units after there size, with the biggeste first
                    stops[x].TransportUnits.Sort((transport1, transport2) =>
                    {
                        if (transport1.UnitType < transport2.UnitType)
                            return 1;
                        if (transport1.UnitType > transport2.UnitType)
                            return -1;
                        return 0;
                    });

                    // Moves transport units to extra route until the trailer is full
                    for(int i = 0; i < stops[x].TransportUnits.Count; )
                    {
                        TransportUnit transportUnit = stops[x].TransportUnits[i];
                        if (transportUnit.UnitType + newStop.GetSizeOfTransportUnits() < extraRoute.GetCapacity())
                        {
                            stops[x].TransportUnits.RemoveAt(0);
                            newStop.TransportUnits.Add(transportUnit);
                        }
                        else
                        {
                            i += 1;
                        }
                    }

                    // Adds the extra route to the total collection of routes
                    Routes.Add(extraRoute);
                }
            }

            // Returns the best load.
            return removedStops;
        }


        /// <summary>
        /// Gives a percentage of completion as an int of 0-100 for status of optmization.
        /// </summary>
        /// <returns>Int value from 0-100</returns>
        public int GetStatus()
        {
            try
            {
                //Calculates count1
                int max1 = overloadedRoutes.Count;
                double steps1 = count1 / max1;
                double status1 = Math.Floor(steps1 * 100) / 2;

                //Calculates count2
                int max2 = removedStops.Count;
                double steps2 = count2 / max2;
                double status2 = Math.Floor(steps2 * 100) / 2;

                //Calculates total.
                double total = status1 + status2;
                return Convert.ToInt32(total);
            }
            catch (NullReferenceException) { return 0; }
            catch (DivideByZeroException) { return 0; }
        }

    }
}
